ddcommon-ffi/src/slice_mut.rs

```diff
@@ -0,0 +1,295 @@
+#[derive(Copy, Clone)]
+pub struct MutSlice<'a, T: 'a> {
+    /// Should be non-null and suitably aligned for the underlying type. It is
+    /// allowed but not recommended for the pointer to be null when the len is
+    /// zero.
+    ptr: *mut T,
+
+    /// The number of elements (not bytes) that `.ptr` points to. Must be less
+    /// than or equal to [isize::MAX].
+    len: usize,
+    _marker: PhantomData<&'a mut [T]>,
+}
+
+impl<'a, T: 'a> core::ops::Deref for MutSlice<'a, T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl<T: Debug> Debug for MutSlice<'_, T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.as_slice().fmt(f)
+    }
+}
+
+impl<T: Eq> PartialEq<Self> for MutSlice<'_, T> {
+    fn eq(&self, other: &Self) -> bool {
+        **self == **other
+    }
+}
+
+impl<T: Eq> Eq for MutSlice<'_, T> {}
+
+/// Use to represent strings -- should be valid UTF-8.
+pub type CharMutSlice<'a> = MutSlice<'a, c_char>;
+
+/// Use to represent bytes -- does not need to be valid UTF-8.
+pub type ByteMutSlice<'a> = MutSlice<'a, u8>;
+
+/// This exists as an intrinsic, but it is private.
+pub fn is_aligned_and_not_null<T>(ptr: *const T) -> bool {
+    !ptr.is_null() && is_aligned(ptr)
+}
+
+#[inline]
+fn is_aligned<T>(ptr: *const T) -> bool {
+    debug_assert!(!ptr.is_null());
+    ptr as usize % std::mem::align_of::<T>() == 0
+}
+
+pub trait AsBytes<'a> {
+    fn as_bytes(&self) -> &'a [u8];
+
+    #[inline]
+    fn try_to_utf8(&self) -> Result<&'a str, Utf8Error> {
+        std::str::from_utf8(self.as_bytes())
+    }
+
+    fn try_to_string(&self) -> Result<String, Utf8Error> {
+        Ok(self.try_to_utf8()?.to_string())
+    }
+
+    #[inline]
+    fn try_to_string_option(&self) -> Result<Option<String>, Utf8Error> {
+        Ok(Some(self.try_to_string()?).filter(|x| !x.is_empty()))
+    }
+
+    #[inline]
+    fn to_utf8_lossy(&self) -> Cow<'a, str> {
+        String::from_utf8_lossy(self.as_bytes())
+    }
+
+    #[inline]
+    /// # Safety
+    /// Must only be used when the underlying data was already confirmed to be utf8.
+    unsafe fn assume_utf8(&self) -> &'a str {
+        std::str::from_utf8_unchecked(self.as_bytes())
+    }
+}
+
+impl<'a> AsBytes<'a> for MutSlice<'a, u8> {
+    fn as_bytes(&self) -> &'a [u8] {
+        self.as_slice()
+    }
+}
+
+impl<'a> AsBytes<'a> for MutSlice<'a, i8> {
+    fn as_bytes(&self) -> &'a [u8] {
+        // SAFETY: safe to convert *i8 to *u8 and then read it... I think.
+        unsafe { MutSlice::from_raw_parts(self.ptr.cast(), self.len) }.as_mut_slice()
+    }
+}
+
+impl<'a, T: 'a> MutSlice<'a, T> {
+    /// Creates a valid empty slice (len=0, ptr is non-null).
+    #[must_use]
+    // TODO: this can't be const under 1.78, but can under later versions.
+    pub fn empty() -> Self {
+        Self {
+            ptr: [].as_mut_ptr(),
+            len: 0,
+            _marker: PhantomData,
+        }
+    }
```

Comment:
You make it const using safe code only with `NonNull::<T>::dangling().as_ptr()` or such.

Reply:
That works for empty.  For now, we still get 
```
error: `std::ptr::NonNull::<T>::new` is not yet stable as a const fn
```

Looks like it went const in 1.85

Reply:
Switched to nonnull, which at least lets empty be const